airflow/models/connection.py

```diff
@@ -109,22 +109,24 @@ def __init__(
             self.extra = extra
 
     def parse_from_uri(self, uri):
-        temp_uri = urlparse(uri)
-        hostname = temp_uri.hostname or ''
-        conn_type = temp_uri.scheme
+        uri_parts = urlparse(uri)
+        hostname = uri_parts.hostname or ''
+        conn_type = uri_parts.scheme
         if conn_type == 'postgresql':
             conn_type = 'postgres'
+        elif '-' in conn_type:
+            conn_type = conn_type.replace('-', '_')
```

Comment:
A general comment @mik-laj .  Should not we rather change the Google Connection scheme to be google-cloud-platform rather than replace '_' with '-' ? I think it's more accurate, simply the google_cloud_platform scheme is wrong - according to https://tools.ietf.org/html/rfc3986#page-17  the scheme part of the URI can only contain alphanumeric characters, hyphen or dot. So urlparse behaves quite well and we should simply change the scheme to be google-cloud-platform

Reply:
In the context of format of  conn_type, this is a permitted character.
In the context of the URL format, this is not allowed.
We can make a conn_type change to be compatible with the url format. but this change can be problematic. In many places `conn_type` with an underscore is already present. I decided that it would be better to provide a greater level of backwards compatibility at the expense of code readability. The current solution does not cause any changes that break compatibility.

Reply:
This implementation requires providing `google-cloud-platform` values as a schema. but after analyzing by urllib.parse, a forbidden character from the schema part is converted into an equivalent. You must provide `google-cloud-platform` to create a connection with conn_type equals to `google_cloud_platform`.

Reply:
Note that no conn_type uses the hyphen character except `'mesos_framework-id`, which does not allow you to read configuration from an environment variable.

Reference:
https://github.com/apache/airflow/blob/f430dbedc5b4c68cdff9d6d760b29bdd73c0682f/airflow/models/connection.py#L54-L91
https://github.com/apache/airflow/blob/c9a82d48a30f09baacb369979ffab6ed39fd02c3/airflow/contrib/executors/mesos_executor.py#L88-L93

Reply:
OK. Make sense! I agree it's the best way of solving it  this way. Good analysis and solution !